### ISSUE TITLE
Proof-of-concept: CSS-only navbar-toggle/navbar-collapse, progressively enhanced with unobtrusive JS

### ISSUE THREAD

**Issue by maniqui** [twbs/bootstrap#15786]
You may know there is a little issue with the "hamburger" menu: currently (as of BS v3.3.2), it won't work without JS. 
Issue has already been reported & discussed in the past (#10506) and closed by @mdo as "won't fix". 

I bring it back to the table with these two examples I've prepared.

### Bootstrap navbar-collapse without Javascript (CSS-only):

* http://jsbin.com/sevula/2/  
(output only; if needed, _resize your window_ so the navbar collapses and the hamburguer icon appears)
* http://jsbin.com/sevula/2/edit?html,css,output

The solution is based on a well-known trick: using the `:target` pseudo-class to apply styles to hide/show an element (you may have seen this trick used for CSS-only modals, for example).

Now, the progressively enhanced version, with unobtrusive Javascript.

### Bootstrap navbar-collapse with unobstrusive Javascript (progressive enhancement)

* http://jsbin.com/duvili/3/  
(output only; if needed, _resize your window_ so the navbar collapses and the hamburguer icon appears)
* http://jsbin.com/duvili/3/edit?html,css,output

This second example is built on top of the CSS-only one: it uses _the same_ CSS. In fact, the CSS used in the CSS-only example isn't even used/applied to this second JS-enhanced version! The HTML for the navbar markup is also the same for both examples. 
The only additions are the necessary `script` tags for adding the proper JS. 

#### So, how does it works?

As stated, the HTML for the navbar remains the same for both versions (CSS-only & JS-enhanced).

The __CSS-only version__ relies on using an anchor (`<a>`) instead of a `<button>`, with the `href` attribute set to the corresponding  `id` of the collapsable navbar element. More precisely:

```html
      <!-- Static navbar -->
      <nav class="navbar navbar-default" id="navbar">
        <div class="container-fluid">
          <div class="navbar-header">
            <a href="#navbar-collapse" class="navbar-toggle collapsed" data-toggle="collapse" aria-expanded="false" aria-controls="navbar">
              <span class="sr-only">Toggle navigation</span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </a>
            <a class="navbar-brand" href="#">Project name</a>
          </div>
          <div id="navbar-collapse" class="navbar-collapse collapse">
            <ul class="nav navbar-nav">
              <li class="active"><a href="#">Home</a></li>
              <li><a href="#">About</a></li>
              <li><a href="#">Contact</a></li>
            </ul>
            <a class="navbar-collapse-close" href="#navbar">Close menu</a>            
          </div><!--/.nav-collapse -->
        </div><!--/.container-fluid -->
      </nav>
```

You will notice that there is also another anchor after the the `div.navbar-nav`. This one is used to create the clickable overlay that will "close" (hide) the `navbar-collapse`. I set is `href` to `#navbar` so to have somewhere to scroll back (and get focus) when the user closes the menu.

The __JS-enhanced version__ uses the collapse.js and transition.js plugins for Bootstrap. I've inlined them because it was the easier way I found to have them included on a JSbin.
And that's all.

#### "What about those "no-js" selectors in the stylesheet? We ain't use that in Bootstrap!"

Those styles are applied to the CSS-only version, or to the JS-enhanced version before JS runs (or if it fails to run). The selectors prefixed and isolated using the good ol' `.no-js` trick, so they won't apply when the class is switched to `.js`.

Disclaimer: I don't currently know how progressive enhancement & unobtrusive JS techniques are approached by Bootstrap developers (that is, I don't know "the Bootstrap way"). 
So for the sake of these examples, and to make them as clear and obvious as possible, I decided to go with the 'no-js/js' route, which is well known by most web-developers. 
I'll appreciate some feedback in the matter.

#### Does it need any changes or hacks to any Bootstrap file?

No, it doesn't. 
It works with Bootstrap v3.3.2 out of the box. Of course, it requires the use of an `<a>` instead of a `<button>`, but both the Bootstrap CSS and the collapse.js plugin already accept the usage of an `<a>` for the "hamburger" icon. 

#### What about accessibility?

I'm not an expert, but I think this approach to navbar-toggle/navbar-collapse _might_ be a bit more accessible than the current way promoted by Bootstrap docs. 
They work _without JS_, which I think it's already an instant win when it comes to accessibility. On top of that, the use of an `<a>` instead of a `<button>` may work as an "skip-to-navbar" anchor (in the CSS-only version and/or when JS fails).

Feedback is welcome.

**Comment by twbs-lmvtfy** [comment]
Hi @maniqui!

You appear to have posted a live example (http://jsbin.com/viforo/3/edit), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 16, column 153: Bad value ```` button ```` for attribute ```` type ```` on element ```` a ````: Subtype missing.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @maniqui!

You appear to have posted a live example (http://jsbin.com/wunipi/1/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by cvrebert** [comment]
Note: We're not going to make any significant changes to the navbar in the rest of Bootstrap v3.x.x, so this would have to target Bootstrap v4, where we're already planning to overhaul the navbar.

**Comment by maniqui** [comment]
@cvrebert: OK, understood, thanks for the feedback.

BTW, let me say that this post is not a feature request, but more a post to show that unobtrusive JS for navbar-collapse is achievable with current Bootstrap (v3.x.x) code, without changes to core files.
 
Web-developers using Bootstrap v3.x.x can just simply follow the suggestions on the opening post (change `<button>` for `<a>`, and add some scoped styles) and get a navbar-collapse that works with & without JS.

**Comment by maniqui** [comment]
Also, I've fixed the complains of that @twbs-lmvtfy robot.

Here are the fixed versions. Now, @twbs-lmvtfy, I demand you to delete your comments! And that's an order! :-P

### Bootstrap navbar-collapse without Javascript (CSS-only):

* http://jsbin.com/sevula/1/  
(output only; if needed, _resize your window_ so the navbar collapses and the hamburguer icon appears)
* http://jsbin.com/sevula/1/edit?html,css,output

### Bootstrap navbar-collapse with unobstrusive Javascript (progressive enhancement)

* http://jsbin.com/duvili/2/  
(output only; if needed, _resize your window_ so the navbar collapses and the hamburguer icon appears)
* http://jsbin.com/duvili/2/edit?html,css,output

**Comment by patrickhlauke** [comment]
Unfortunately, the pure CSS version does not work well for keyboard users (they can open the menu, but not close it anymore unless they follow some link), so this has some accessibility concerns (though granted, still more accessible than having the menu not work at all without JS).

For the unobtrusive JS version, if you are changing from ``<button>`` to ``<a>``, you should also include a ``role="button"`` to expose the correct role to assistive technologies.

**Comment by maniqui** [comment]
> Unfortunately, the pure CSS version does not work well for keyboard users (they can open the menu, but not close it anymore unless they follow some link), so this has some accessibility concerns (though granted, still more accessible than having the menu not work at all without JS).

You are right. They can't close it by just hitting the Intro key. One way to close it (beside following some link to other page) is tabbing to the "Close menu" link (which works, at the same time, as the clickable overlay).

**Comment by patrickhlauke** [comment]
> tabbing to the "Close menu" link

which gives no visual indication of being there unfortunately, so sighted keyboard users will likely only discover it by accident, if at all.

**Comment by maniqui** [comment]
Adding a `:focus` style may help. Will look into it later.

**Comment by maniqui** [comment]
I've done some minor improvements to add styles for `:focus` state on the "Close menu" link, as an aid to sighted keyboard users. 
Also, in the source order, I've moved the "Close menu" link above the `.navbar-nav`, so when navigating using the Tab key (through the navbar), it comes after the "hamburger" icon, and before the `navbar-nav`. I'm just _guessing_ that this could be a more intuitive order...

**Comment by mdo** [comment]
I believe that using the `:target` option also pollutes the browser history quite heavily, no? I tried that out before switching to a checkbox based solution (seen in http://lanyon.getpoole.com). That, for the most part, works pretty well. Unsure about the accessibility of it though.

**Comment by maniqui** [comment]
You are correct: it does pollute the browser history on each click. Though, one would wonder why the user would be frenetically showing/hiding the menu, without clicking on any other link (being it another `#anchor` on the same page or a link to a different URL). Notwithstanding, this history polluting might be an usability/accessibility issue.

Re: the sibling checkbox trick on Lanyon. Nice usage of it

**Comment by malikfaizanhaider** [comment]
Id is not same!

**Comment by maniqui** [comment]
What do you mean @malikfaizanhaider ?

**Comment by mdo** [comment]
Punting on this for now—it's something we could perhaps visit in v4, but not the initial release. Please do consider releasing it as an add-on though!
